### PR TITLE
02 09 family name

### PR DESCRIPTION
**Pull request by jlandwehr-avarteq**
In this exercise I display all family members in people#show. People are automatically associated by their last name and ordered in ASC. Furthermore, the person itself is excluded from the list.

Note: I implemented this functionality with a name scope but had to give person with as an argument, because I was not able to implement it with self. => Please show me how to do it during the review.

@arkanoryn

### CODE REVIEW

File: family/app/models/person.rb [hunk 1]
```diff
@@ -15,4 +20,11 @@ class Person < ActiveRecord::Base
 
   accepts_nested_attributes_for :address
   validates_associated :address
+
+  #searches for all family members, order is not necessary and only there  for exercise purposes
+  #scope :family_members, -> (last_name){ where(last_name: last_name).order('first_name asc') }
+scope :family_members, -> (person){ where(last_name: person.last_name)
+                                    .where.not(first_name: person.first_name)
+                                    .order('first_name asc') }
+  #.order('first_name 'ASC)
```
Comment (arkanoryn) [inline family/app/models/person.rb]: ??

File: family/app/models/person.rb [hunk 2]
```diff
@@ -15,4 +20,11 @@ class Person < ActiveRecord::Base
 
   accepts_nested_attributes_for :address
   validates_associated :address
+
+  #searches for all family members, order is not necessary and only there  for exercise purposes
+  #scope :family_members, -> (last_name){ where(last_name: last_name).order('first_name asc') }
```
Comment (arkanoryn) [inline family/app/models/person.rb]: ??

File: family/app/models/person.rb [hunk 3]
```diff
@@ -15,4 +20,11 @@ class Person < ActiveRecord::Base
 
   accepts_nested_attributes_for :address
   validates_associated :address
+
+  #searches for all family members, order is not necessary and only there  for exercise purposes
```
Comment (arkanoryn) [inline family/app/models/person.rb]: Wrong comment.
`Searches for all family members except the person send...`

File: family/app/models/person.rb [hunk 4]
```diff
@@ -1,6 +1,11 @@
 class Person < ActiveRecord::Base
-  has_one :address,              dependent: :destroy
-  has_many :pets,                dependent: :destroy
+  has_one :address,              as: :owner,    dependent: :destroy
```
Comment (arkanoryn) [inline family/app/models/person.rb:2]: align everything or align nothing.

File: family/app/controllers/degrees_controller.rb [hunk 5]
```diff
@@ -0,0 +1,98 @@
+class DegreesController < ApplicationController
+  before_action :set_degree, only: [:show, :edit, :update, :destroy]
```
Comment (arkanoryn) [inline family/app/controllers/degrees_controller.rb]: you know how to use a better way for your Array.

File: family/app/controllers/degrees_controller.rb [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  # GET /degrees
+  # GET /degrees.json
+  def index
+    @degrees = Degree.all
+  end
+
+  # GET /degrees/1
+  # GET /degrees/1.json
+  def show
+
```
Comment (arkanoryn) [inline family/app/controllers/degrees_controller.rb]: empty line?

File: family/app/controllers/degrees_controller.rb [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  end
+
+  # GET /degrees/new
+  def new
+    @degree = Degree.new
+  end
+
+  # GET /degrees/1/edit
+  def edit
+
```
Comment (arkanoryn) [inline family/app/controllers/degrees_controller.rb]: empty line?

File: family/app/controllers/people_controller.rb [hunk 8]
```diff
@@ -1,6 +1,8 @@
 class PeopleController < ApplicationController
   before_action :set_person, only: [:show, :edit, :update, :destroy]
```
Comment (arkanoryn) [inline family/app/controllers/people_controller.rb]: you know a better way to declare arrays.

File: family/app/controllers/people_controller.rb [hunk 9]
```diff
@@ -1,6 +1,8 @@
...
-  before_action :set_animal_types, only: [:show]
+  before_action :set_animal_types, only: :show
+  before_action :set_degrees, only: [:new, :edit, :create, :update]
```
Comment (arkanoryn) [inline family/app/controllers/people_controller.rb]: idem

File: family/app/controllers/people_controller.rb [hunk 10]
```diff
@@ -1,6 +1,8 @@
...
+  before_action :select_possible_friends, only: [:edit, :update]
```
Comment (arkanoryn) [inline family/app/controllers/people_controller.rb]: idem.

File: family/app/controllers/people_controller.rb [hunk 11]
```diff
@@ -11,6 +13,8 @@ def index
   # GET /people/1
   # GET /people/1.json
   def show
+    #@family = Person.family_members(@person.last_name)
```
Comment (arkanoryn) [inline family/app/controllers/people_controller.rb]: what is this comment for?

File: family/app/controllers/people_controller.rb [hunk 12]
```diff
@@ -42,6 +47,9 @@ def create
   # PATCH/PUT /people/1
   # PATCH/PUT /people/1.json
   def update
+    set_person_degrees_from_params
+    set_person_friends_from_params
```
Comment (arkanoryn) [inline family/app/controllers/people_controller.rb]: what?!

File: family/app/controllers/people_controller.rb [hunk 13]
```diff
@@ -93,4 +110,14 @@ def person_params
       ]
     )
   end
+
+  #selects all people that are not friends with @person
+  def select_possible_friends
+    @possible_friends = Person.all.to_a - @person.friends.to_a
```
Comment (arkanoryn) [inline family/app/controllers/people_controller.rb]: could you not make `@possible_friends = Person.all.to_a - @person.friends.to_a - [@person]`?

File: family/app/controllers/schools_controller.rb [hunk 14]
```diff
@@ -0,0 +1,84 @@
+class SchoolsController < ApplicationController
+  before_action :set_school, only: [:show, :edit, :update, :destroy]
```
Comment (arkanoryn) [inline family/app/controllers/schools_controller.rb]: can do better

File: family/app/controllers/schools_controller.rb [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+  def destroy
+    @school.destroy
+    respond_to do |format|
+      format.html { redirect_to schools_url, notice: 'School was successfully destroyed.' }
+      format.json { head :no_content }
+    end
+  end
+
+  private
+    # Use callbacks to share common setup or constraints between actions.
```
Comment (arkanoryn) [inline family/app/controllers/schools_controller.rb:67]: space before/after private

File: family/app/models/degree.rb [hunk 16]
```diff
@@ -0,0 +1,8 @@
+class Degree < ActiveRecord::Base
+  has_many :people_degrees ,                dependent: :destroy
```
Comment (arkanoryn) [inline family/app/models/degree.rb]: alignment?

File: family/app/models/friendship.rb [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+  after_create :create_inverse, unless: :inverse_exists?
+
+  # After destroying a Friendship, I want to destroy the inverses, if inverses exists
+  after_destroy :destroy_inverse, if: :inverse_exists?
+
+  private
+
+  # Create the inverse Friendship
+  def create_inverse
+    #Friendship.create(inverse_friendship)
```
Comment (arkanoryn) [inline family/app/models/friendship.rb]: ??

File: family/app/models/school.rb [hunk 18]
```diff
@@ -0,0 +1,11 @@
+class School < ActiveRecord::Base
+  has_one  :address, as: :owner, dependent: :destroy
+  has_many :school_degrees,    dependent: :destroy
```
Comment (arkanoryn) [inline family/app/models/school.rb]: ??

File: family/app/models/school.rb [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+  has_many :degrees, through: :school_degrees
+  has_many :people, through: :degrees
+
+  validates :school_name, presence: true
```
Comment (arkanoryn) [inline family/app/models/school.rb:7]: ??

File: family/app/views/degrees/show.html.erb [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+
+<p>
+  <strong>Title:</strong>
+  <%= @degree.title %>
+</p>
+
+<p><strong>Schools:</strong></p>
+  <% @degree.schools.each do |school| %>
+    <p>
+      <%= school.school_name %>
```
Comment (arkanoryn) [inline family/app/views/degrees/show.html.erb:11]: ??

**Review by arkanoryn — CHANGES_REQUESTED** [review]
I gave you some changes to do.
Please, apply and merge on `master`.

**Comment by jlandwehr-avarteq** [comment]
Merge done.

**Pull request closed without merging**